### PR TITLE
Improve Cypress runner command line config

### PR DESCRIPTION
**Pull request by nemanjaglumac**
### Status
PENDING REVIEW

### What does this PR accomplish?
- Improves the way we build `commandLineConfig` in Cypress runner, which opens up the possibility to extend it easier in the future
- Implements [the official Cypress recommendation](https://docs.cypress.io/guides/references/configuration#Command-Line) on how to pass the complex value to the `--config` flag (JSON.stringified object)

### Additional context
We're programatically building the CLI command for spinning up Cypress. This PR is concerned with `commandLineConfig` in particular. Instead of relying on string concatenation and `if-else-if` logic, it feels much more natural to simply build the config object from self-contained pieces.

*For one of the next steps we will need to pass the array to `testFiles`. That wasn't possible before with the string literals approach.*

Let's say we want to run the specs from the "question" folder only. The command to run is:
`yarn test-cypress-open --spec frontend/test/metabase/scenarios/question/**/*.cy.spec.js`

And the resulting config would look like this:
```bash
$ /Users/username/path/to/metabase/node_modules/.bin/cypress open
  --config-file frontend/test/cypress.json
  --config '{"baseUrl":"http://localhost:4000","testFiles":"frontend/test/metabase/scenarios/question/**/*.cy.spec.js"}' 
  --env HAS_ENTERPRISE_TOKEN=true
```

Opening Cypress without additional flags (`yarn test-cypress-open`) results in the following config:
```bash
$ /Users/username/path/to/metabase/node_modules/.bin/cypress open
  --config-file frontend/test/cypress.json
  --config '{"baseUrl":"http://localhost:4000","ignoreTestFiles":"**/metabase-{smoketest,db}/**"}'
  --env HAS_ENTERPRISE_TOKEN=true
```

### Links:
- https://docs.cypress.io/guides/guides/command-line
- https://docs.cypress.io/guides/references/configuration

### CODE REVIEW

**Comment by nemanjaglumac** [comment]
@ariya this is the first step towards manual grouping of Cypress tests. I will build on top of this in the next PR, and I'd like to suggest that we use `--spec` exclusively. It's more flexible than the `--folder` flag that changes the whole `integrationFolder`.

With just a simple update to how we parse `userArgs` it will be possible to point Cypress to multiple arbitrary folders without physically changing them (rearranging). We might end up rearranging test folders and specs within them for different reasons at some point in the future, but this feels like the good first step that works with what we currently have.

To sum it up, you'd be able to run a command like:
`yarn test-cypress-open --spec "path/to/specific/foo.spec.js,path/to/specific/bar.spec.js" (which doesn't have real world application)

What we can use for grouping is the fact that `--spec` accepts globs
```
yarn test-cypress-open
  --group queries
  --spec "frontend/test/metabase/scenarios/question/**/*.cy.spec.js,frontend/test/metabase/scenarios/dashboard/**/*.cy.spec.js"
```

```
yarn test-cypress-open
  --group visualizations
  --spec "frontend/test/metabase/scenarios/visualizations/**/*.cy.spec.js"
```

File: frontend/test/__runner__/run_cypress_tests.js [hunk 1]
```diff
@@ -73,15 +73,26 @@ const init = async () => {
...
+    folderConfig || specsConfig
+      ? null
+      : { ignoreTestFiles: "**/metabase-{smoketest,db}/**" };
+
+  const config = {
+    ...baseConfig,
+    ...folderConfig,
+    ...specsConfig,
+    ...ignoreConfig,
+  };
```
Comment (daltojohnso) [inline frontend/test/__runner__/run_cypress_tests.js:92]: `JSON.stringify` removes `undefined` properties (not `null` though), so you can do something like this to make things a little more compact/readable:

```
const config = {
  baseUrl: server.host,
  integrationFolder: sourceFolder ? sourceFolderLocation : undefined,
  testFiles: singleFile ? singleFileName : undefined,
  ignoreTestFiles: (sourceFolder || singleFile) ? undefined : '**/metabase-{smoketest,db}/**'
};
```

**Review by daltojohnso — APPROVED** [review]

File: frontend/test/__runner__/run_cypress_tests.js (hunk 1 above)
Comment (nemanjaglumac) [reply frontend/test/__runner__/run_cypress_tests.js:92]: Neat!
I didn't know that it completely removes `undefined` but doesn't remove `null`. Thanks.

While I agree that this is definitely more succinct, I'm afraid it can become harder to read (especially if we start adding more key/value pairs and more conditional logic).

**Review by alxnddr — APPROVED** [review]

**Review by kulyk — APPROVED** [review]
